jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/message/MessageInputStream.java

```diff
@@ -105,11 +94,14 @@ public void appendFrame(ByteBuffer framePayload, boolean fin) throws IOException
                         state = State.SUSPENDED;
                         break;
 
-                    case SUSPENDED:
+                    default:
                         throw new IllegalStateException();
                 }
 
-                buffers.put(copy);
+                // Put the payload into the queue, by copying it.
+                // Copying is necessary because the payload will
+                // be processed after this method returns.
+                buffers.put(copy(framePayload));
```

Comment:
We should probably be paying attention to the `fin` parameter on this `appendMessage(ByteBuffer, boolean)` and add the EOF sentinel to the buffers if it's set to true.